### PR TITLE
Allow user-defined Javascript to run after hot reload

### PR DESCRIPTION
**Pull request by fieldstrength**
This patch allows the user to specify the javascript they want to run every time the page finishes updating the content with a hot-reload.

They do so by defining the function named `emaHotReloadHook()` in their custom `<head>`.

## Motivation 

My neuron/emanote notebook relies heavily on MathJax support to render mathematics. Upon migrating to Emanote, I needed a way to instruct MathJax to redraw it. MathJax provides instructions for handling such a case [here](https://docs.mathjax.org/en/latest/advanced/typeset.html#mathjax-in-dynamic-content).

In principle the way Ema replaces `<script>` tags might have taken care of this case. However it seemed not to work, as the initial hot reload event that arrives right after opening the page replaced the rendered math (visible only for a flash) with the unprocessed source text. 

In addition, there are known problems pertaining to multiple concurrent math renders. Since the first hot-reload event arrives very quickly after the page loads, it is important to have direct control over the JS invoked upon reload events to deal gracefully with this.

## Alternatives considered

- I initially suggested that Ema should accept configuration (i.e. in Haskell) that contains the desired javascript directly. In many ways I find such a direct and explicit approach favorable, however it would seemingly require rather significant changes to the Haskell type signatures to accommodate (there is no place to simple add a field or similar). Given that this is a minor feature, and given the limited valuable time of the contributors for this open source project, the more simple solution implemented in this patch may be the better tradeoff.

- We also considered indicating through some special attributes that a particular `<script>` tag was to be run on hot reload. However these approaches also had problems
  - If the script has type `text/javascript` then they are always run when the page first loads in addition to after the reload events, which defeats the purpose of the feature. For my use-case it is critical to have code that runs *only* after reloads to properly deal with race conditions involving the initial render.
  - If the script has some other type, like `text/ema-hot-reload` then, as desired, the script does not run on page load. But we'd need to use something like `eval` to run it. `eval` is generally considered undesirable both from a security perspective, as well as for potentially interfering with browser Javascript optimizations. Overall this approach also has more complexity compared to the simple solution.

### CODE REVIEW

**Comment by fieldstrength** [comment]
In case it helps anyone else, here is the relevant head configuration Im currently using with my Emanote zettelkasten to allow MathJax to work rather smoothly. In particular you can see how Im defining `emaHotReloadHook`.

```yaml
  headHtml: |
    <link href="https://cdn.jsdelivr.net/npm/prismjs@1.23.0/themes/prism-tomorrow.css" rel="stylesheet" />
    <script src="https://cdn.jsdelivr.net/combine/npm/prismjs@1.23.0/prism.min.js,npm/prismjs@1.23.0/plugins/autoloader/prism-autoloader.min.js"></script>
    <script>
      window.mathjaxInitialTypesetDone = false;
      window.MathJax = {
        startup: {
          ready: () => {
            MathJax.startup.defaultReady();
            MathJax.startup.promise.then(() => {
              console.log('MathJax typesetting complete');
              window.mathjaxInitialTypesetDone = true;
            });
          }
        }
      };
    </script>
    <script async="" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>
      function emaHotReloadHook() {
        if (window.mathjaxInitialTypesetDone) { MathJax.typeset(); }
      }
    </script>
```

**Comment by srid** [comment]
@fieldstrength Can you share the Markdown that will allow me to reproduce the original bug? Because using your Math include but without the `emaHotReloadHook` - I cannot reproduce the original bug with say `$a + b ^ 2 = c$`.

File: src/Ema/Server.hs [hunk 1]
```diff
@@ -354,6 +354,9 @@ wsClientShim =
             } else {
               console.log("ema: ✍ Patching DOM");
               setHtml(document.documentElement, evt.data);
+              if ("emaHotReloadHook" in window) {
+                emaHotReloadHook();
```
Comment (srid) [inline src/Ema/Server.hs:358]: You want a `console.log` here, so that we can use browser console log for latter debugging at runtime.

**Comment by srid** [comment]
https://github.com/srid/emanote/commit/2a4d87afeacd6b11bd185a83b8096aa5c53e183d#diff-9c54c38cbcab94236ffc9f41e85777a1cf27f342b6c43ba08636c89a5b6b0a49R1-R37

That works on live server for me.

**Comment by srid** [comment]
On the mermaid page however I see this on first access (no issue for 2nd+ accesses):

image

(with `master` branch)

**Comment by fieldstrength** [comment]
Well, this is very unexpected, but you're right @srid. Something in that MathJax config here is mostly fixing the issue for me too. I didn't notice because that was not at all my motivation for writing that config.

It's still not perfect; Subsequent hot reload events cause light breakage maybe about 1/4 of the time. But Im finding it's the same situation with my solution using this patch.

I leave it to you to decide if we should just close this PR for now. It seems something like this might be needed eventually, but I'd understand not wanting to include it until its necessary and demonstrably useful. 

If you want to start by documenting what we've found, here is a more compact [startup config](http://docs.mathjax.org/en/latest/options/startup/startup.html#the-configuration-block) for MathJax that achieves this "mostly" solution:

```html
    <script>
      window.MathJax = {
        startup: {
          ready: () => {
            MathJax.startup.defaultReady();
          }
        }
      };
    </script>
    <script async="" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
```

**Comment by fieldstrength** [comment]
Closing this for now, since we found another solution to the motivating problem

**Pull request closed without merging**